### PR TITLE
Infer nuget package sources for nuget.config

### PR DESCRIPTION
**Pull request by AnthonySteele**
#278 

Infer nuget package sources from the file `nuget.config` when it is present and not overridden on the commandline.

The fallback sequence is now:
 * If you specify nuget sources on the command line options, they will be used. User override wins.
 * if a `nuget.config` file is found and can be parsed for sources, those sources will be used.
 * The global feed on `nuget.org` is the fallback option. This will actually be 100% correct for many public projects, and will give some meaningful results for even internal code.

The new classes start with `NugetSourcesReader` that does the above logic.  And reader classes called by that to parse the file.

Most of the changes are because the `NuGetSources`  cannot now be determined from config and then put into the IoC container upfront. Instead, when there is a Folder containing the code, it does the `NugetSourcesReader` logic, and thereafter passes along the `NuGetSources` object. `NuGetSources` is no longer a constructor param, now it's a method param. This touches a lot of classes [in order to to pass it to where it is used](https://en.wiktionary.org/wiki/tramp_data).

### CODE REVIEW

File: NuKeeper.Inspection.Tests/Sources/NugetConfigFileParserTests.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        [Test]
+        public void ValidDataWithOneItemReturnsThatItem()
+        {
+            var data =
+@"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+  <packageSources>
+    <add key=""NuGet official"" value=""https://nuget.org/api/v2/"" />
```
Comment (skolima) [inline NuKeeper.Inspection.Tests/Sources/NugetConfigFileParserTests.cs]: I know it's the default at your $work, and this is just test data, but it's a bit odd to see v2 here :D

File: NuKeeper.Inspection.Tests/Sources/NugetSourcesReaderTests.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            var ff = new FolderFactory(new NullNuKeeperLogger());
+            return ff.UniqueTemporaryFolder();
+        }
+
+        private static INugetSourcesReader MakeNuGetSourcesReader()
+        {
+            var logger = new NullNuKeeperLogger();
+            return new NugetSourcesReader(
+                new NugetConfigFileReader
+                    (new NugetConfigFileParser(logger), logger), logger);
```
Comment (skolima) [inline NuKeeper.Inspection.Tests/Sources/NugetSourcesReaderTests.cs]: 😂

File: NuKeeper.Inspection.Tests/Sources/NugetConfigFileParserTests.cs (hunk 1 above)
Comment (AnthonySteele) [reply NuKeeper.Inspection.Tests/Sources/NugetConfigFileParserTests.cs]: Yep, i should change that.

File: NuKeeper.Inspection.Tests/Sources/NugetSourcesReaderTests.cs (hunk 2 above)
Comment (AnthonySteele) [reply NuKeeper.Inspection.Tests/Sources/NugetSourcesReaderTests.cs]: I wanted to refactor, but then I decided to keep the tests on the `NugetConfigFileParser` so I could not

**Comment by AnthonySteele** [comment]
Yep, we can't really traverse the whole file system.

**Comment by skolima** [comment]
Ha, you've  edited your comment so my late-night revelation isn't a revelation any more, but: came to the same conclusion :-) I was thinking about the scenarios "full" NuGet deals with, a developer having multiple levels of the `nuget.config` in their work folder and arranging projects in there - but for NuKeeper scenario, the only concern really is  a top level configuration file in the repository root.

**Comment by AnthonySteele** [comment]
Yep, this should work as a repo-level config file; either in the root folder or sometimes one level down, e.g. in a `/src` folder.

**Pull request merged**